### PR TITLE
Half angle fresnel

### PR DESCRIPTION
**Pull request by raytranuk**
This PR:
- is built on top of the envBRDF PR (https://github.com/playcanvas/engine/pull/3157) as it didn't make sense to revert the envBRDF changes
- switches punctual lights to use half-angle fresnel
- removes some code complexity dealing with differences to where area lights calculate fresnel

Below are some screen captures before and after this PR (Tip: its best to open each image in a separate tab and switch tabs to see the difference):

Before:
image
After:
image

Before:
image
After:
image

As requested, here is a list of shader chunks altered by this PR:
- combineClearCoat.frag 
- combineDiffuseSpecular.frag 
- combineDiffuseSpecularNoConserve.frag 
- combineDiffuseSpecularNoRefl.frag 
- combineDiffuseSpecularNoReflSeparateAmbient.frag 
- combineDiffuseSpecularOld.frag 
- fresnelSchlick.frag 
- lightSpecularAnisoGGX.frag 
- lightSpecularBlinn.frag 
- lightSpecularPhong.frag 
- ltc.frag

I confirm I have read the [contributing guidelines](https://github.com/playcanvas/engine/blob/master/.github/CONTRIBUTING.md) and signed the [Contributor License Agreement](https://docs.google.com/a/playcanvas.com/forms/d/1Ih69zQfJG-QDLIEpHr6CsaAs6fPORNOVnMv5nuo0cjk/viewform).

### CODE REVIEW

**Comment by Maksims** [comment]
Just as a note: advanced users do override shader chunks for custom materials, e.g. water shader.
Often engine changes to such chunks, makes Devs work harder, as they need to migrate often.

I'm aware that chunks are subject to change, and is not a public API. Just writing it to ensure that changes and really necessary, as they will affect advanced users.

Also, we need a concrete long-term solution for custom shader chunks. Nodes are probably an option here. Although they are more towards tech artists, not shader programmers.

**Comment by raytranuk** [comment]
> Just as a note: advanced users do override shader chunks for custom materials, e.g. water shader.
> Often engine changes to such chunks, makes Devs work harder, as they need to migrate often.
> 
> I'm aware that chunks are subject to change, and is not a public API. Just writing it to ensure that changes and really necessary, as they will affect advanced users.

@Maksims - thanks for the note - it is still under discussion whether half angle fresnel for punctual lights should be added as an option with new shaders chunks or replace existing lighting chunks - I'll keep this PR as draft until that discussion is resolved.

> Also, we need a concrete long-term solution for custom shader chunks. Nodes are probably an option here. Although they are more towards tech artists, not shader programmers.

You may have seen announced that there is a node based shader editor being developed - however this will, at least initially, affect lighting.

File: src/graphics/program-lib/programs/standard.js [hunk 1]
```diff
@@ -1493,14 +1488,9 @@ var standard = {
                 code += "   addReflection();\n";
             }
 
-            if (options.useSpecular && options.useMetalness && options.fresnelModel > 0) code += "   getFresnel();\n";
+            // NB switched to more correct half-angle fresnel
```
Comment (mvaligursky) [inline src/graphics/program-lib/programs/standard.js:1491]: any reason to have 'NB' on many of the comments?

**Review by mvaligursky — APPROVED** [review]
looks good as far as I can tell.

File: src/graphics/program-lib/programs/standard.js (hunk 1 above)
Comment (raytranuk) [reply src/graphics/program-lib/programs/standard.js:1491]: For this one its a note for future users who may wonder why some logic was removed - its true that some NBs can be removed

**Comment by slimbuck** [comment]
Looks good! @raytranuk could you please add to the PR description a list of shader chunks that the PR changes?

**Review by slimbuck — APPROVED** [review]

**Pull request merged**